MDEV-28671 Enable var deprecation for mysqld help output

christian7007 opened the pull request:
## Description

Currently there are mechanisms to mark a system variable as deprecated, but it is only used to print warning messages when a deprecated variable is set.

The goal of this PR is to leverage the existing mechanisms in order to make the deprecation information available at the `--help` output of `mysqld`. The following modifications have been added:

* Moving the deprecation information (i.e `deprecation_substitute`
attribute) from the `sys_var` class into the `my_option` struct.
As every `sys_var` contains its own `my_option` struct, the access
to the deprecation information remains available to `sys_var`
objects. `my_getotp` functions, which works directly with
`my_option` structs, gain access to this information while building
the `--verbose --help` output.

* Change the `option_cmp` function to use the `deprecation_substitute`
attribute (when it's set) instead of the name when sorting the options.
This way deprecated options and its substitutes are grouped together.

* For plugin variables, leverages the `PLUGIN_VAR_DEPRECATED` flag
and set the `deprecation_substitute` attribute accordingly when
building the `my_option` objects.

The above changes affects the `--verbose --help` output the following way:

- When a system variable `deprecation_substitute` is set to `OPT_DEPRECATED_NO_REPLACEMENT`:
   - If there's a comment for the variable, the `(DEPRECATED)` tag will appear along with the comment:
      ```
      --big-tables        (DEPRECATED)
                      Old variable, which if set to 1, allows large result sets
                      by saving all temporary sets to disk, avoiding 'table
                      full' errors. No longer needed, as the server now handles
                      this automatically.
      ```
   - If there's **not** a comment for the variable, the variable will not be printed. 

- When a system variable `deprecation_substitute` is set to something different than `OPT_DEPRECATED_NO_REPLACEMENT`:
   - If there's a comment for the variable, the `(DEPRECATED and replaced by '<string>')` tag will appear along with the comment. This comment should be used for adding extra information on how to use the substitute variable.
      ```
      --wsrep-causal-reads (DEPRECATED and replaced by 'wsrep_sync_wait')
                      Use wsrep_sync_wait READ flag instead.
      --wsrep-sync-wait[=#]
                      Ensure "synchronous" read view before executing an
                      operation of the type specified by bitmask: 1 -
                      READ(includes SELECT, SHOW and BEGIN/START TRANSACTION);
                      2 - UPDATE and DELETE; 4 - INSERT and REPLACE
      ```
   -  If there's **not** a comment for the variable, the `(DEPRECATED and replaced by '<string>')` tag will appear next to the variable name. The variable will appear right above the substitute variable, so it will be easier for the user to read the comment available for the substitute.
      ```
      --old-alter-table[=name] (DEPRECATED and replaced by 'alter_algorithm')
      --alter-algorithm[=name]
                      Specify the alter table algorithm. One of: DEFAULT, COPY,
                      INPLACE, NOCOPY, INSTANT
      ```
-  When a plugin variable have the `PLUGIN_VAR_DEPRECATED` flat set the `(DEPRECATED)` tag will appear next to the name.

### Future work
Review existing variables and ensure the deprecation mechanisms are properly used. If this proposal is accepted, I can create a follow-up PR going through existing variables and marking them as deprecated (i.e set `deprecation_substitute`) accordingly.

## How can this PR be tested?
The corresponding MTR test has been added, so just running the MTR tests as below should be enough (any other test should work as before):

```
./mtr --suite main --parallel=auto
```

## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch*

## Copyright
All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=2376) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=2376) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=2376) it.</sub>

grooverdan commented on mysys/my_getopt.c:
```diff
@@ -1586,6 +1588,22 @@ void my_print_help(const struct my_option *options)
+      if (strcmp(optp->deprecation_substitute, OPT_DEPRECATED_NO_REPLACEMENT) == 0){
+        col= print_comment("(DEPRECATED)",
+                           col, name_space, comment_space);
+      }
```
style: no braces for a single like in an `if`/`else` statement.

grooverdan reviewed (commented):
Nice overall. I could use the additional comment field in #2357. Please look at test errors.

christian7007 commented:
Thanks for the quick review @grooverdan! I've updated the PR fixing an issue at `fill_sysvars`, which failed with a `SIGSEV` when the comment was not set, and a few minor issues (MTRs result files and `deprecation_substitute` initializations). It seems that tests are passing now, let me know if you need me to add/change anything else.

christian7007 commented:
I have just rebased the PR agains `10.11`, please let me know if you need me to change the target branch to the latest `11.0`.

grooverdan commented:
Just leave as is for the moment until reviewed. Based on target versions in the MDEV it *may* end up earlier too. But for now, please just wait for the review and rest assured that the intention is to get it reviewed by the next release at the end of the month.

vuvova commented on mysys/my_getopt.c:
```diff
@@ -1546,7 +1546,9 @@ void my_print_help(const struct my_option *options)
-    if (!optp->comment)
+    if (!optp->comment &&
+        (optp->deprecation_substitute == NULL ||
+          strcmp(optp->deprecation_substitute, OPT_DEPRECATED_NO_REPLACEMENT) == 0))
```
why you didn't use the existing convention that empty string (`""`) means no replacement?

vuvova replied:
why does it matter here whether the option is deprecated or not? no comment means it's invisible, why would deprecation suddenly make it visible again?

christian7007 replied:
> why you didn't use the existing convention that empty string ("") means no replacement?

My goal was just to make the code more self explanatory, reading `OPT_DEPRECATED_NO_REPLACEMENT` will provide more information than `""`.

> why does it matter here whether the option is deprecated or not? no comment means it's invisible, why would deprecation suddenly make it visible again?

The idea behind it is that when a variable is just deprecated by another one, the `deprecation_substitute` will be set and the description field can be then re-used to add extra information on how to achieve the same behaviour, if needed. So if an option has been just renamed or directly replaced by another one, the `deprecation_substitute` will be set and the comment will be emptied, resulting in something like:

```
--old-variable[=name] (DEPRECATED and replaced by 'new-variable')
--new-variable[=name]  new-variable description
```

But if for example, in order to achieve the same behaviour we need to do something more specific we can add that in the comment filed, and should result in something like:

```
--old-variable[=name] (DEPRECATED and replaced by 'new-variable')
                       Use `X` value of new-variable to achieve the same behaviour.
--new-variable[=name]  new-variable description
```

vuvova replied:
you can make the code more self explanatory without `strcmp` and `strdup`. That is, keep the existing convention where `""` means no replacement, and add something like
```c
#define DEPRECATED_NO_REPLACEMENT(X) (*(X) == 0)
```
Then you can write
```c
    if (!optp->comment &&
        (optp->deprecation_substitute == NULL ||
          DEPRECATED_NO_REPLACEMENT(optp->deprecation_substitute)))
```
which is arguably even more self explanatory than `strcmp`

vuvova replied:
Before your changes the option with no comment was not printed by `--help`. It was an intentional feature, please, don't break it.

vuvova commented on mysys/my_getopt.c:
```diff
@@ -1586,6 +1588,21 @@ void my_print_help(const struct my_option *options)
+        col= print_comment("')",
+                           col, name_space, comment_space);
+      }
+    }
```
may be closer to what the server was printed before? And why are you SHOUTING?
`(deprecated)` and `(deprecated, use %s instead)` would be better

vuvova replied:
I don't see  a warning when a deprecated option is used. Like in `mysqld --debug`

christian7007 replied:
> I don't see a warning when a deprecated option is used. Like in mysqld --debug

I didn't think about that part, the current PR only add the logic to print the deprecation message for the `mysqld --help --verbose` output and when the variable is set via `SET var_name`. But I think we should add that too, if you agree, for the next revision I can add the logic for printing the warning at the bottom of:
https://github.com/MariaDB/server/blob/c37ebaf6c224e7ba571653eecd48a080d4c1a061/sql/mysqld.cc#L7967

vuvova commented on sql/set_var.cc:
```diff
@@ -1186,8 +1182,27 @@ int fill_sysvars(THD *thd, TABLE_LIST *tables, COND *cond)
+        fields[7]->store(buf, buflen, scs);
+        free(buf);
+      }
+    }
```
I'm not sure about it. `--help` appends lots of various stuff to the comment. Like "Enabled by default use --skip to disable" or "One of: option1, option2, option3", etc. And now "(disabled, use %s instead)". None of the other such --help auto-additions are appended to the comment in INFORMATION_SCHEMA.SYSTEM_VARIABLES.

I don't think the disabled comment should be an exception. Better add a separate column to the I_S table.

vuvova commented on sql/sql_plugin.cc:
```diff
@@ -4003,6 +4003,13 @@ static int construct_options(MEM_ROOT *mem_root, struct st_plugin_int *tmp,
+    {
+      options->deprecation_substitute= (char*) alloc_root(mem_root,
+                                                          sizeof(OPT_DEPRECATED_NO_REPLACEMENT));
+      options->deprecation_substitute= OPT_DEPRECATED_NO_REPLACEMENT;
```
again, why would you need to create a copy of a string here?

vuvova commented on mysql-test/main/innodb_mysql_sync.result:
```diff
@@ -455,7 +455,7 @@ info: Records: 0  Duplicates: 0  Warnings: 0
 affected rows: 0
 Warnings:
-Warning	1287	'@@old_alter_table' is deprecated and will be removed in a future release. Please use '@@alter_algorithm' instead
+Warning	1287	'@@old_alter_table' is deprecated and will be removed in a future release. Please use alter_algorithm instead
```
looks rather wrong. Please, restore the old message

vuvova commented on sql/sys_vars.cc:
```diff
@@ -2597,11 +2597,11 @@ static Sys_var_max_user_conn Sys_max_user_connections(
 static Sys_var_ulong Sys_max_tmp_tables(
-       "max_tmp_tables", "Unused, will be removed.",
+       "max_tmp_tables", NULL,
        SESSION_VAR(max_tmp_tables), CMD_LINE(REQUIRED_ARG),
```
I think the old help text made sense. Someone looking at the config file might want to know what `max_tmp_table=10000` did.

vuvova commented on mysql-test/main/mysqld--help.result:
```diff
@@ -13,6 +13,7 @@ The following specify which files/extra groups are read (specified before remain
  xxx_deinit(). That also means that one can load any
  function from any library, for example exit() from
  libc.so
+ --old-alter-table[=name] (DEPRECATED and replaced by 'alter_algorithm')
```
Ideally should be `by --alter-algorithm`

christian7007 commented:
I've just updated the PR with the requested changes, this is the summary of the changes:

- Replace `sub_size` by `deprecation_substitute` in the `my_option` struct instead of adding new field.
- Keep using `""` as the value for deprecated variables without replacement, and use macros to make the code more self explanatory as suggested [here](https://github.com/MariaDB/server/pull/2376#discussion_r1072934020).
- Skip printing any option (either deprecated or not) if comment is `null`.
- Adjust the deprecation messages to the ones used by the server before (i.e `ER_WARN_DEPRECATED_SYNTAX_NO_REPLACEMENT` and `ER_WARN_DEPRECATED_SYNTAX`).

   ```
   # mysqld --verbose --help output

   --old-alter-table[=name] (it's deprecated and will be removed in a future
                      release. Please use '--alter-algorithm' instead)
                      ...
  --date-format=name  (it's deprecated and will be removed in a future release)
                      ...
   
   # SET VAR warnings

    MariaDB [(none)]> SET max_tmp_tables=100;
    Query OK, 0 rows affected, 1 warning (0.000 sec)
    MariaDB [(none)]> show warnings;
    +---------+------+--------------------------------------------------------------------------+
    | Level   | Code | Message                                                                  |
    +---------+------+--------------------------------------------------------------------------+
    | Warning | 1287 | '@@max_tmp_tables' is deprecated and will be removed in a future release |
    +---------+------+--------------------------------------------------------------------------+
    1 row in set (0.000 sec)
    
    MariaDB [(none)]> SET old_alter_table=COPY;
    Query OK, 0 rows affected, 1 warning (0.000 sec)
    
    MariaDB [(none)]> show warnings;
    +---------+------+-----------------------------------------------------------------------------------------------------------------+
    | Level   | Code | Message                                                                                                         |
    +---------+------+-----------------------------------------------------------------------------------------------------------------+
    | Warning | 1287 | '@@old_alter_table' is deprecated and will be removed in a future release. Please use @@alter_algorithm instead |
    +---------+------+-----------------------------------------------------------------------------------------------------------------+
    1 row in set (0.000 sec)
   ```

- Print a warning in the log when a deprecated argument/option is used (note it won't print any warning with `--debug` as this is not marked as deprecated, check `Future work` section on PR description).

   ```
   $ ./sql/mysqld --max-tmp-table=100
    2023-01-20 21:13:01 0 [Note] Using unique option prefix 'max-tmp-table' is error-prone and can break in the future. Please use the full name 'max_tmp_tables' instead.
    2023-01-20 21:13:01 0 [Warning] 'max_tmp_tables' is deprecated and will be removed in a future release
    2023-01-20 21:13:01 0 [Note] Starting MariaDB 10.11.2-MariaDB-debug source revision 8eefd345e9b96124321209623b9101e9b93f8c21 as process 46920
    ...
    
   $ ./sql/mysqld --old-alter-table=COPY
    2023-01-20 21:13:56 0 [Warning] 'old_alter_table' is deprecated and will be removed in a future release. Please use --alter-algorithm instead
    2023-01-20 21:13:56 0 [Note] Starting MariaDB 10.11.2-MariaDB-debug source revision 8eefd345e9b96124321209623b9101e9b93f8c21 as process 47592
   ...
   ```
- Remove deprecation information from the `INFORMATION_SCHEMA.SYSTEM_VARIABLES` comment field.

**NOTE**: Many of the MTRs changes just remove single quotes wrapping variable names. These quotes were added before because the `deprecation_substitute` was defined like `deprecation_substitute="'@@var_name'"`. IMHO we should just store the name of the variable in `deprecation_substitute` so we can easily format it depending on the context (i.e add `@@` or `--`, replace `-` by `_`, ...). Also, if quotes are preferred in the logs we can add them directly in the error message here:
https://github.com/MariaDB/server/blob/26ef4875e6ba17cb1cf83e4eec5b39b6ab5f89e8/sql/share/errmsg-utf8.txt#L5412

christian7007 commented:
`buildbot/amd64-debian-10-deb-autobake` check is failing because of a timeout. Every other checks, including other debian10 checks, seem to pass properly, so I'm not sure if the error is related with my changes.

christian7007 commented:
I have just updated the PR addressing the conflicts with the target branch.

LinuxJedi reviewed (approved):
I think this is good now, but we would need to rebase it on 11.0. I defer to @vuvova to make sure he is happy that his concerns are resolved before merging.

christian7007 commented:
I've rebased my changes a couple of times with the latest changes on `11.0` branch having the CI failing with different errors which seems unrelated to my changes. Looking at the [history](https://github.com/MariaDB/server/commits/11.0) of commits in the branch it looks like the CI is in an unstable state right now. Please let me know if you think the errors might be related with my changes, and I'll try to dig deeper on it.

LinuxJedi commented:
> I've rebased my changes a couple of times with the latest changes on `11.0` branch having the CI failing with different errors which seems unrelated to my changes. Looking at the [history](https://github.com/MariaDB/server/commits/11.0) of commits in the branch it looks like the CI is in an unstable state right now. Please let me know if you think the errors might be related with my changes, and I'll try to dig deeper on it.

Yes, at the moment a patch has broken the ColumnStore engine compiling against 11.0. The ColumnStore team are aware and are working on fixes. In the mean time it is causing a couple of builders to fail.

vuvova commented on mysys/my_getopt.c:
```diff
@@ -1588,6 +1588,26 @@ void my_print_help(const struct my_option *options)
+        strxmov(buf1, "--", optp->deprecation_substitute, NullS);
+        convert_underscore_to_dash(buf1, strlen(optp->deprecation_substitute) + 2);
+
+        col= print_comment("(it's deprecated and will be removed in a future release. Please use '",
```
And, please, add tests for setting a deprecated variable. An easy way to do that would be to change `--enable-cleartext-plugin` in `mysql.cc` to use the new deprecated api. Than you can add, like `exec $MYSQL -enable-cleartext-plugin` to `mysql.test`

christian7007 replied:
Waiting for discussion [here](https://github.com/MariaDB/server/pull/2376#discussion_r1203442851) before implement it. We would need to add support for the `deprecation_substitue` into `mysql` client.

christian7007 commented on mysql-test/main/alter_table.result:
```diff
@@ -1677,7 +1677,7 @@ ALTER TABLE t1 DROP INDEX i1, DROP INDEX i2, DROP INDEX i3, DROP INDEX i4;
 affected rows: 0
 Warnings:
-Warning	1287	'@@old_alter_table' is deprecated and will be removed in a future release. Please use '@@alter_algorithm' instead
+Warning	1287	'@@old_alter_table' is deprecated and will be removed in a future release. Please use @@alter_algorithm instead
```
Ok, I'll update the error message here as I said in my previous comment then:
https://github.com/MariaDB/server/blob/26ef4875e6ba17cb1cf83e4eec5b39b6ab5f89e8/sql/share/errmsg-utf8.txt#L5412

christian7007 commented on sql/mysqld.cc:
```diff
@@ -8436,6 +8459,21 @@ mysqld_get_one_option(const struct my_option *opt, const char *argument,
+      sql_print_warning("\'%s\' is deprecated and will be removed in a future release. Please use %s instead", opt->name, buf1);
+    }
+  }
+
```
I might be missing something but I didn't put this within my_getopt as this is used for both system variables and command line arguments, so if we print the warning there I won't be able to know if I have to print it using the variable format (i.e using `@@` and `_`) or the argument format (i.e using `--` and `-`).

vuvova replied:
What do you mean? This part of `mysqld.cc` is only used for command-line options. Deprecation warnings for `SET` is printed in `set_var.cc`

christian7007 commented on sql/mysqld.cc:
```diff
@@ -7691,6 +7707,13 @@ static int option_cmp(my_option *a, my_option *b)
+
+  if (deprecated_b)
+    return 1;
+
```
The idea behind this is to print the deprecated variables grouped with the corresponding deprecation substitute, so if deprecation substitute is set we'll use that within the sort function (instead of the variable name) to group them together.

christian7007 commented:
Sorry for the late replay, I think I have address all the feedback, but the comments I have replied to. This is a quick summary of the main changes:

- Added quotes for ER_WARN_DEPRECATED_SYNTAX at server/sql/share/errmsg-utf8.txt.
- Add `=1` for `wsrep_sync_wait` again.
- Adjust both `convert_dash_to_underscore()` and `convert_underscore_to_dash` to stop when the first alphanumeric value other than `-` or `_` respectively is found.

NOTE: I updated the branch last week, it seems the CI is stuck and I'm still waiting for the results.

christian7007 commented:
I'm closing this PR as I won't be able to keep working on it. But I don't mind if anyone else want to iterate on it and finish it.

vuvova commented:
Thanks! I'll try to apply it then.
